### PR TITLE
[13.0][ADD] account_payment_fee_transfer

### PR DESCRIPTION
**Pull request by Saran440**
This module extends the functionality of the
Payment Transfer to put fee, manage journal entry and support multi currency

Example : Internal Transfer Bank1 -> Bank2 (default, no fee)
```
        Account             | Debit     | Credit
        =========================================================
        Journal Entry 1
        ---------------
        BANK1               |           |   1000.0
        TRANSFER            |   1000.0  |

        Journal Entry 2
        ---------------
        TRANSFER            |           |   1000.0
        BANK2               |   1000.0  |
```

Example : Internal Transfer Bank1 -> Bank2 (fee on BANK1)

```
        Account             | Debit     | Credit
        =========================================================
        Journal Entry 1
        ---------------
        BANK1               |           |   1000.0
        BANK1               |           |     10.0        --> Fee
        ACCOUNT FEE         |     10.0  |
        TRANSFER            |   1000.0  |

        Journal Entry 2
        ---------------
        TRANSFER            |           |   1000.0
        BANK2               |   1000.0  |
```
Example main company is currency `EUR`
For case : Internal Transfer Bank1 (EUR) -> Bank3 (USD) (fee on BANK1)
```
        Account             | Debit     | Credit     | Amount Currency
        ===============================================================
        Journal Entry 1
        ---------------
        BANK1               |           |   1000.0   |
        BANK1               |           |     10.0   |                --> Fee
        ACCOUNT FEE         |     10.0  |            |
        TRANSFER            |   1000.0  |            |

        Journal Entry 2
        ---------------
        TRANSFER            |           |   1000.0   |
        BANK3               |   1000.0  |            |   1283.4
```

For case : Internal Transfer Bank1 (EUR) -> Bank3 (USD) (fee on BANK3)
```
        Account             | Debit     | Credit     | Amount Currency
        ===============================================================
        Journal Entry 1
        ---------------
        BANK1               |           |   1000.0   |
        TRANSFER            |   1000.0  |            |

        Journal Entry 2
        ---------------
        TRANSFER            |           |   1000.0   |
        ACCOUNT FEE         |     10.0  |            |
        BANK3               |           |     10.0   |    -12.83       --> Fee
        BANK3               |   1000.0  |            |   1283.40
```

issue : https://github.com/OCA/account-payment/issues/305

### CODE REVIEW

File: account_payment_fee_transfer/models/account_payment.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                        payment.fee_transfer,
+                        company_currency,
+                        payment.company_id,
+                        payment.payment_date,
+                    )
+                    currency_id = payment.currency_id.id
+
+                # Add fee transfer in journal deduct.
+                if journal_id == payment.deduct_journal_id.id:
+                    credit_account = (
```
Comment (kittiu) [inline account_payment_fee_transfer/models/account_payment.py]: As this is ID not record
```suggestion
                    credit_account_id = (
```

File: account_payment_fee_transfer/models/account_payment.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                        payment.deduct_journal_id.default_debit_account_id.id
```
Comment (kittiu) [inline account_payment_fee_transfer/models/account_payment.py]: As we get for credit account. A little more inline.
```suggestion
                        payment.deduct_journal_id.default_credit_account_id.id
```

File: account_payment_fee_transfer/models/account_payment.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+from odoo import fields, models
+
+
+class AccountPayment(models.Model):
+    _inherit = "account.payment"
+
+    fee_transfer = fields.Monetary(
+        readonly=True, states={"draft": [("readonly", False)]}, copy=False
+    )
```
Comment (kittiu) [inline account_payment_fee_transfer/models/account_payment.py:16]: help="Fee being charged into the account of 'Deduct From'"

File: account_payment_fee_transfer/readme/DESCRIPTION.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+This module extends the functionality of the Payment Transfer to put fee
```
Comment (kittiu) [inline account_payment_fee_transfer/readme/DESCRIPTION.rst]:
```suggestion
This module extends the functionality of the Payment Transfer by allowing to set the bank charge (fee) amount to either source or destination bank account.
```

File: account_payment_fee_transfer/readme/USAGE.rst [hunk 5]
```diff
@@ -0,0 +1,5 @@
+To use this module, you need to:
+
+#. Go to Invoicing > Customers / Vendors > Payments
+#. Create new payment and select payment type = 'transfer'
+#. fill fee in 'fee transfer' field and details.
```
Comment (kittiu) [inline account_payment_fee_transfer/readme/USAGE.rst]:
```suggestion
#. Fill amount for bank fee in 'fee transfer' field and select the bank charge account and bank account the fee incurred.
```

**Comment by kittiu** [comment]
Testing with negative amount, please add constraints
Selection_234

**Comment by kittiu** [comment]
@wipawii can you help check the journal entry and user experience?
Selection_237
Selection_236

**Comment by kittiu** [comment]
@OCA/accounting-maintainers 
This module simple allow additional bank charge to the standard account.payment's internal transfer. But I think the name of module should be account_payment_transfer_bank_fee. Any suggests?

**Comment by wipawii** [comment]
@kittiu This sample looks like the company currency base is USD. If so, The accounting entries is correct.

The concept of cash/bank transfer account move is
- From A > Credit A
- To B > Debit B
- Bank charge (if any)
  - Debit Fee Account (Expense bank charge)
  - Credit Deduct from (Account)

Hope this helps.

I would ask about if the company currency base is not USD, let say, it is THB and there is USD bank account, accounting should record gain/loss on exchange rate for this transfer. **Shall we add this case in?**

**Comment by kittiu** [comment]
@Saran440 please add for case different currency of bank.
Note: the gain loss exchange rate account should already exists some where already, you can use it. You can test gain loss from invoice in USD and payment in THB.

**Comment by Saran440** [comment]
@kittiu Fixed. Thank you for your reviewed.
Example main company is currency `EUR`
For case : Internal Transfer Bank1 (EUR) -> Bank3 (USD) (fee on BANK1)
```
        Account             | Debit     | Credit     | Amount Currency
        ===============================================================
        Journal Entry 1
        ---------------
        BANK1               |           |   1000.0   |
        BANK1               |           |     10.0   |                --> Fee
        ACCOUNT FEE         |     10.0  |            |
        TRANSFER            |   1000.0  |            |

        Journal Entry 2
        ---------------
        TRANSFER            |           |   1000.0   |
        BANK3               |   1000.0  |            |   1283.4
```

For case : Internal Transfer Bank1 (EUR) -> Bank3 (USD) (fee on BANK3)
```
        Account             | Debit     | Credit     | Amount Currency
        ===============================================================
        Journal Entry 1
        ---------------
        BANK1               |           |   1000.0   |
        TRANSFER            |   1000.0  |            |

        Journal Entry 2
        ---------------
        TRANSFER            |           |   1000.0   |
        ACCOUNT FEE         |     10.0  |            |
        BANK3               |           |     10.0   |    -12.83       --> Fee
        BANK3               |   1000.0  |            |   1283.40
```

**Comment by kittiu** [comment]
@OCA/accounting-maintainers 
for this case, I am not very sure it is useful for everyone.
I would like to invite for some comments, whether this is good to merge, and also its name.
If this one is ok, we also have the batch transfer for it https://github.com/OCA/account-payment/pull/317

If not objection, we will continue the merge in a while. :)

**Pull request closed without merging**